timemachine/fe/single_topology_v3.py

```diff
@@ -735,3 +742,182 @@ def setup_intermediate_state(self, lamb):
+
+        combined_torsion = potentials.PeriodicTorsion(combined_torsion_idxs).bind(combined_torsion_params)
+
+        # concatenate ligand charges with host charges
+        ligand_exclusions = []
+        ligand_scale_factors = []
+
+        num_ligand_atoms = self.get_num_atoms()
+
+        for i in range(num_ligand_atoms):
+            for j in range(i + 1, num_ligand_atoms):
+                ligand_exclusions.append((i, j))
+                ligand_scale_factors.append((1.0, 1.0))
+
+        ligand_exclusions = np.array(ligand_exclusions, dtype=np.int32) + num_host_atoms
+        ligand_scale_factors = np.array(ligand_scale_factors, dtype=np.float64)
+
+        combined_exclusion_idxs = np.concatenate([host_nonbonded.get_exclusion_idxs(), ligand_exclusions])
+        combined_scale_factors = np.concatenate([host_nonbonded.get_scale_factors(), ligand_scale_factors])
+
+        host_qlj = host_nonbonded.params
+
+        ligand_charges = []
+        ligand_sigmas = []
+        ligand_epsilons = []
+        ligand_lambda_plane_idxs = []
+        ligand_lambda_offset_idxs = []
+
+        # generate charges and lj parameters for each ligand
+        ligand_a_q = self.ff.q_handle.parameterize(self.mol_a)
+        ligand_a_lj = self.ff.lj_handle.parameterize(self.mol_a)
+
+        ligand_b_q = self.ff.q_handle.parameterize(self.mol_b)
+        ligand_b_lj = self.ff.lj_handle.parameterize(self.mol_b)
+
+        c_to_a = dict()
+        for k, v in enumerate(self.a_to_c):
+            c_to_a[v] = k
+
+        c_to_b = dict()
+        for k, v in enumerate(self.b_to_c):
+            c_to_b[v] = k
+
+        for idx, membership in enumerate(self.c_flags):
+            if membership == 0:  # core atom
+                a_idx = c_to_a[idx]
+                b_idx = c_to_b[idx]
+                plane = 0
+                offset = 0
+                # interpolate charges when in common-core
+                q = (1 - lamb) * ligand_a_q[a_idx] + lamb * ligand_b_q[b_idx]
+                sig = (1 - lamb) * ligand_a_lj[a_idx, 0] + lamb * ligand_b_lj[b_idx, 0]
+                eps = (1 - lamb) * ligand_a_lj[a_idx, 1] + lamb * ligand_b_lj[b_idx, 1]
+            elif membership == 1:  # dummy_A
+                a_idx = c_to_a[idx]
+                plane = 0
+                offset = 1
+                q = ligand_a_q[a_idx]
+                sig = ligand_a_lj[a_idx, 0]
+                eps = ligand_a_lj[a_idx, 1]
+            elif membership == 2:  # dummy_B
+                b_idx = c_to_b[idx]
+                plane = -1
+                offset = 1
+                q = ligand_b_q[b_idx]
+                sig = ligand_b_lj[b_idx, 0]
+                eps = ligand_b_lj[b_idx, 1]
+            else:
+                assert 0
+
+            ligand_lambda_plane_idxs.append(plane)
+            ligand_lambda_offset_idxs.append(offset)
+            ligand_charges.append(q)
+            ligand_sigmas.append(sig)
+            ligand_epsilons.append(eps)
```

Comment:
Organization: Would it be useful to extract this stretch into a different function `self.combine_params(a_qlj, b_qlj) -> (c_qlj, c_plane_idxs, c_offset_idxs)` or something, since it contains the main "opinionated" steps (e.g. how to interpolate core q, sig, eps)? May also be good to test this separately.

(The concatenation logic that occupies the rest of the function seems fairly closed-ended.)

Reply:
I think that's a good idea - @mcwitt would that be useful for you?